lib/internal/bootstrap/node.js

```diff
@@ -301,33 +301,47 @@ function startup() {
   } = perf.constants;
   perf.markMilestone(NODE_PERFORMANCE_MILESTONE_BOOTSTRAP_COMPLETE);
 
-  return startExecution;
+  if (isMainThread) {
+    return startMainThreadExecution;
+  } else {
+    return startWorkerThreadExecution;
+  }
+}
+
+function startWorkerThreadExecution() {
+  prepareUserCodeExecution();
+
+  // If we are in a worker thread, execute the script sent through the
+  // message port.
+  const {
+    getEnvMessagePort,
+    threadId
+  } = internalBinding('worker');
+  const {
+    createMessageHandler,
+    createWorkerFatalExeception
+  } = NativeModule.require('internal/process/worker_thread_only');
+
+  // Set up the message port and start listening
+  const debug = NativeModule.require('util').debuglog('worker');
+  debug(`[${threadId}] is setting up worker child environment`);
+
+  const port = getEnvMessagePort();
+  port.on('message', createMessageHandler(port));
+  port.start();
+
+  // Overwrite fatalException
+  process._fatalException = createWorkerFatalExeception(port);
 }
 
 // There are various modes that Node can run in. The most common two
 // are running from a script and running the REPL - but there are a few
 // others like the debugger or running --eval arguments. Here we decide
 // which mode we run in.
-function startExecution() {
-  // This means we are in a Worker context, and any script execution
-  // will be directed by the worker module.
-  if (!isMainThread) {
-    const workerThreadSetup = NativeModule.require(
-      'internal/process/worker_thread_only'
-    );
-    // Set up the message port and start listening
-    const { workerFatalExeception } = workerThreadSetup.setup();
-    // Overwrite fatalException
-    process._fatalException = workerFatalExeception;
-    return;
-  }
-
-  // To allow people to extend Node in different ways, this hook allows
-  // one to drop a file lib/_third_party_main.js into the build
-  // directory which will be executed instead of Node's normal loading.
-  if (NativeModule.exists('_third_party_main')) {
+function startMainThreadExecution(mainScript) {
+  if (mainScript) {
```

Comment:
This bit here is temporary - I think it would be better to do the branching in C++ and directly `compileAndCall` the main scripts instead. For per-process CLI options like `--help` we could branch in `node::Start()` since they do not make a lot of sense for either embedders or the workers anyway. Although unfortunately for `_third_party_main` we have to keep a driver that wrap it into a `process.nextTick()` (or maybe there is a cleverer way out?)